### PR TITLE
drivers: eeprom: Add MB85RC FRAM driver 

### PR DESCRIPTION
**Pull request by msierszulski**
This PR adds support for MB85RC FRAM driver.

### CODE REVIEW

File: drivers/eeprom/eeprom_mb85rc.c [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+	return rem;
+}
+
+static int mb85rc_init(const struct device *dev)
+{
+	const struct mb85rc_config *cfg = dev->config;
+	struct mb85rc_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+
```
Comment (henrikbrixandersen) [inline drivers/eeprom/eeprom_mb85rc.c:78]: You need to check if the I2C controller device is ready and error out if not.

File: dts/bindings/mtd/fujitsu,mb85rc.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2023 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: Fujitsu MB85RC series I2C FRAM
+
+compatible: "fujitsu,mb85rc"
```
Comment (henrikbrixandersen) [inline dts/bindings/mtd/fujitsu,mb85rc.yaml:6]: Should this include placeholder x's? As in `fujitsu,mb85rcxx`? I couldn't find an existing devicetree binding in the Linux kernel for this.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Please add this to `tests/drivers/build_all/eeprom/`. Otherwise looks good!

File: drivers/eeprom/eeprom_mb85rc.c (hunk 1 above)
Comment (msierszulski) [reply drivers/eeprom/eeprom_mb85rc.c]: This is now changed

File: dts/bindings/mtd/fujitsu,mb85rc.yaml (hunk 2 above)
Comment (msierszulski) [reply dts/bindings/mtd/fujitsu,mb85rc.yaml]: Thanks for your suggestion!

**Comment by msierszulski** [comment]
@henrikbrixandersen We have added `mb85rcxx` to `tests/drivers/build_all/eeprom/`.

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
@msierszulski You need to fix the device driver initialization order to make CI pass.

File: drivers/eeprom/Kconfig.mb85rcxx [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+	select I2C
+	help
+	  Enable FUJITSU mb85rcxx i2c FRAM
+
+config EEPROM_MB85RCXX_INIT_PRIORITY
+	int "MB85RCXX FRAM init priority"
+	default 80
+	depends on EEPROM_MB85RCXX
+	help
+	  MB85RCXX FRAM driver device initialization priority
```
Comment (henrikbrixandersen) [inline drivers/eeprom/Kconfig.mb85rcxx:17]: Please do not introduce a driver-specific initialization priority Kconfig. Use `CONFIG_EEPROM_INIT_PRIORIT`Y` but adjust the test, if needed, to ensure the test GPIO controller is initialized before the FRAM driver.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/eeprom/Kconfig.mb85rcxx (hunk 3 above)
Comment (msierszulski) [reply drivers/eeprom/Kconfig.mb85rcxx:17]: Changing the default `CONFIG_EEPROM_INIT_PRIORITY` in `tests/drivers/build_all/eeprom/prj.conf` will be a good solution here?

Comment (henrikbrixandersen) [reply drivers/eeprom/Kconfig.mb85rcxx]: I am unsure why this is even needed. The existing `tests/drivers/build_all/eeprom/` also depends on GPIOs, and it does not fail in priority checking.

The relevant priorities, when running on `native_posix`, are as follows:
```
$ grep -E '(EEPROM|GPIO)_INIT_PRIO' zephyr/.config 
CONFIG_EEPROM_INIT_PRIORITY=50
CONFIG_GPIO_INIT_PRIORITY=40
```

Comment (msierszulski) [reply drivers/eeprom/Kconfig.mb85rcxx]: @henrikbrixandersen The problem appeared in the CI build: https://github.com/zephyrproject-rtos/zephyr/actions/runs/5976393051/job/16215196325?pr=61653#step:13:1266
But I don't know why, I am not able to reproduce it locally. 
For `em_starterkit_em7d` `tests/drivers/build_all/eeprom/drivers.eeprom.build` passes and the priorities are as follows:
```
grep -E '(EEPROM|GPIO)_INIT_PRIO' build/zephyr/.config
CONFIG_EEPROM_INIT_PRIORITY=50
CONFIG_GPIO_INIT_PRIORITY=70
```
Building with `west` or testing with `twister` works for me without error.

Comment (henrikbrixandersen) [reply drivers/eeprom/Kconfig.mb85rcxx]: Ahh, that is another issue. It is solved here: https://github.com/zephyrproject-rtos/zephyr/pull/61951

Comment (msierszulski) [reply drivers/eeprom/Kconfig.mb85rcxx]: Great, we will wait until it is merged to main

Comment (fabiobaltieri) [reply drivers/eeprom/Kconfig.mb85rcxx]: Heya, merged, should be fine to drop this now.

Comment (msierszulski) [reply drivers/eeprom/Kconfig.mb85rcxx]: Thanks!
@henrikbrixandersen We've dropped the priority config changes.

**Review by henrikbrixandersen — DISMISSED** [review]

File: drivers/eeprom/eeprom_mb85rcxx.c [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+static int mb85rcxx_read(const struct device *dev, off_t offset, void *buf, size_t len)
+{
+	const struct mb85rcxx_config *cfg = dev->config;
+	struct mb85rcxx_data *data = dev->data;
+	uint8_t addr[2];
+	uint16_t i2c_addr = 0;
+	size_t len_in_page = 0;
+	int ret = 0;
```
Comment (fabiobaltieri) [inline drivers/eeprom/eeprom_mb85rcxx.c:107]: These don't need to be initialized, drop the `= 0`

File: drivers/eeprom/eeprom_mb85rcxx.c [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+static int mb85rcxx_write(const struct device *dev, off_t offset, const void *buf, size_t len)
+{
+	const struct mb85rcxx_config *cfg = dev->config;
+	struct mb85rcxx_data *data = dev->data;
+	uint8_t addr[2];
+	uint16_t i2c_addr = 0;
+	size_t len_in_page = 0;
+	int ret = 0;
```
Comment (fabiobaltieri) [inline drivers/eeprom/eeprom_mb85rcxx.c:161]: same

File: drivers/eeprom/eeprom_mb85rcxx.c (hunk 4 above)
Comment (msierszulski) [reply drivers/eeprom/eeprom_mb85rcxx.c]: Thanks, I removed the `= 0`

File: drivers/eeprom/eeprom_mb85rcxx.c (hunk 5 above)
Comment (msierszulski) [reply drivers/eeprom/eeprom_mb85rcxx.c]: Thanks, I removed the `=0`

**Review by fabiobaltieri — DISMISSED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by msierszulski** [comment]
@fabiobaltieri @henrikbrixandersen I had to restore initialization of one variable due to failed twister tests. PTAL

**Comment by fabiobaltieri** [comment]
> @fabiobaltieri @henrikbrixandersen I had to restore initialization of one variable due to failed twister tests. PTAL

You could have an explicit "return 0" at the end of that function instead, that would be preferable IMO.

**Review by fabiobaltieri — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**